witness.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2025 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tessera
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"golang.org/x/mod/sumdb/note"
+)
+
+// policyComponent describes a component that makes up a policy. This is either a
+// single Witness, or a WitnessGroup.
+type policyComponent interface {
+	// Satisfied returns true if the checkpoint is signed by the quorum of
+	// witnesses involved in this policy component.
+	Satisfied(cp []byte) bool
+
+	// URLs returns the URLs for requesting a counter signature from all
+	// witnesses that are involved in determining the satisfaction of this
+	// PolicyComponent.
+	URLs() []*url.URL
+}
+
+// NewWitness returns a Witness given a verifier key and the root URL for where this
+// witness can be reached.
+func NewWitness(vkey string, witnessRoot *url.URL) (Witness, error) {
+	v, err := note.NewVerifier(vkey)
+	if err != nil {
+		return Witness{}, err
+	}
+	// "key hash" MUST be a lowercase hex-encoded SHA-256 hash of a 32-byte Ed25519 public key.
+	// This expression cuts off the identity name and hash.
+	key64 := strings.SplitAfterN(vkey, "+", 3)[2]
+	key, err := base64.StdEncoding.DecodeString(key64)
+	if err != nil {
+		return Witness{}, err
+	}
+	h := sha256.Sum256(key)
+
+	u := witnessRoot.JoinPath(fmt.Sprintf("/%x/add", h))
```

Comment:
Should this be `/add-checkpoint`, from https://github.com/C2SP/C2SP/blob/main/tlog-witness.md#add-checkpoint?

Reply:
Good spot. I'd already got this in the follow-up PR that I'm still working on, but I appreciate the attention to detail: https://github.com/transparency-dev/trillian-tessera/compare/main...mhutchinson:trillian-tessera:b309-witnessConfig?expand=1